src/api/erc20/Erc20ApiProxy.ts

```diff
@@ -0,0 +1,58 @@
+import BN from 'bn.js'
+
+import { Receipt } from 'types'
+
+import {
+  Erc20Api,
+  NameParams,
+  SymbolParams,
+  DecimalsParams,
+  TotalSupplyParams,
+  BalanceOfParams,
+  AllowanceParams,
+  ApproveParams,
+  TransferParams,
+  TransferFromParams,
+} from './Erc20Api'
+import { CacheProxy } from 'api/proxy'
+
+const DEFAULT_CACHE_TIME = 30
+
+export class Erc20ApiProxy extends CacheProxy<Erc20Api> implements Erc20Api {
+  public name(params: NameParams): Promise<string> {
```

Comment:
Would it be better, instead of rewriting each function to use `this.fetchWithCache` inside, to accept an params object in constructor with a list of functions to cache and then in constructor:
```
for (const fname of toCache) {
  this[fname] = cacheHOF(api[fname])
}
```

same can be done with TS for types

Reply:
That looks neat!
Yes, this is something I would pretty much would like to have.
The reason for re-implementing all methods is due to the fact that I'm implementing the api interface and TS forced me to implement everything.
Will give this a try.

Reply:
Didn't quite get it working...

Pushed my attempt so far to https://github.com/gnosis/dex-react/commit/9855180f9fa46b5041a323807fb66ce3d0345f89

The proxy implementation doesn't recognize the class as having the methods implemented:

```typescript
class TestApiProxy extends CacheProxy<TestApi> implements TestApi {
  public constructor(api: TestApi) {
    super(api, [{ methodType: 'cachedMethod', expiration: 10 }])
  }

  // public async cachedMethod(params: Params): Promise<string> {
  //   return this.fetchWithCache('cachedMethod', params, 10)
  // }
  // public async nonCachedMethod(params: Params): Promise<string> {
  //   return this.fetchWithCache('nonCachedMethod', params)
  // }
}
```

TS warning: 
```
Class 'TestApiProxy' incorrectly implements interface 'TestApi'.
  Type 'TestApiProxy' is missing the following properties from type 'TestApi': cachedMethod, nonCachedMethod
```

Reply:
Managed to get TS happy in a way that we don't have to re-implement all methods. Take a look at the latest revision.